### PR TITLE
Simplify build file

### PR DESCRIPTION
**Pull request by 3flex**
Simplify the build file a bit by updating the way Kotlin API & language versions are declared, and how the JVM target is set.

### CODE REVIEW

File: build.gradle.kts [hunk 1]
```diff
@@ -198,6 +188,17 @@ kotlin.target.compilations.configureEach {
 }
 
 tasks {
+    withType<KotlinCompilationTask<*>>().configureEach {
```
Comment (TWiStErRob) [inline build.gradle.kts:169]: Why did you move to a global withType, rather than configuring the compilation's options/task?

(Compilations are the way forward, we don't need to know about tasks any more; similar to how AGP's variant and artifact DSL hides the existence of tasks.)

Comment (TWiStErRob) [reply build.gradle.kts:169]: As noted in the removed comment (in commit 2) -- now that kotlin-dsl is not messing with us -- we can use the cleanest DSL:

```kotlin
kotlin.target.compilations.configureEach {
  compilations.compilerOptions.configure {
    ...
  }
}
```
No need for any task references.

File: build.gradle.kts [hunk 2]
```diff
@@ -165,34 +165,18 @@ sourceSets {
...
-
-kotlin.target.compilations.configureEach {
-    compileTaskProvider.configure {
-        // Validate that we're using the right version.
-        doFirst {
-            val api = compilerOptions.apiVersion.get()
-            val language = compilerOptions.languageVersion.get()
-            if (api != usedKotlinVersion || language != usedKotlinVersion) {
-                TODO(
-                    "There's mismatch between configured and actual versions:\n" +
```
Comment (TWiStErRob) [inline build.gradle.kts:198]: 🤞 you're right that no-one else is touching this now, neither in the future (as we maintain deps and evolve).

File: build.gradle.kts [hunk 3]
```diff
@@ -99,9 +98,8 @@ dependencies {
     testImplementation(gradleApi())
 }
 
-java {
-    sourceCompatibility = JavaVersion.VERSION_1_8
-    targetCompatibility = JavaVersion.VERSION_1_8
+kotlin {
+    jvmToolchain(8)
```
Comment (TWiStErRob) [inline build.gradle.kts:103]: While this is nicer, it has a big side effect: many machines need to download a separate JDK just to compile 8. Even JDK 21 supports targeting 8!

Comment (TWiStErRob) [reply build.gradle.kts:103]: I guess stutter downloads them anyway (when running tests), so it's fine.

**Review by TWiStErRob — CHANGES_REQUESTED** [review]

File: build.gradle.kts (hunk 3 above)
Comment (szpak) [reply build.gradle.kts:103]: > I guess stutter downloads them anyway (when running tests), so it's fine.

However, not everyone runs the tests with stutter for JDK 8. In #298 I propose a concept to build (and run tests) with the current version by default. Only the regression tests require other JDK, but it is handled by CI (or can be triggered by a developer, if needed). In that context the original assignment seems more flexible to a regular developer build.

Comment (TWiStErRob) [reply build.gradle.kts:103]: True, we could also argue that since the foojay resolver is pretty common in Gradle builds, the availability of all relevant JDKs is likely on any dev machine doing Gradle-based dev. Even if not, it's an almost negligible one-time cost. If this was actively maintained by hundreds or had cost-sensitive uncached CI, my view would be different.

---

That said, it's probably better to use the default JVM with target 8 as before, so that the target version is not *implicit* from the toolchain. It also reduces any potential contributor's setup complexity.

So the question is what we optimise for? We seem to have a consensus for DX with @szpak. @3flex I'm curious what's are your thoughts on the above points.

File: build.gradle.kts (hunk 1 above)
Comment (3flex) [reply build.gradle.kts:169]: > Why did you move to a global withType

Because [that's what the documentation says to do to](https://kotlinlang.org/docs/gradle-compiler-options.html#target-the-jvm) to modify compiler options. Actually it should be `withType<KotlinJvmCompile>()` because we're not in a multiplatform project, I can update.

The actual cleanest config for setting compiler options would be:
```kotlin
kotlin {
    compilerOptions {
        apiVersion = ...
        languageVersion = ...
    }
}
```

But this requires the project to update to Kotlin 1.9.

There's no mention anywhere [here ](https://kotlinlang.org/docs/gradle-configure-project.html) or [here](https://kotlinlang.org/docs/gradle-compiler-options.html) of the syntax that's currently used or what you proposed above?

File: build.gradle.kts (hunk 3 above)
Comment (3flex) [reply build.gradle.kts:103]: > so that the target version is not implicit from the toolchain

That's a feature, not a bug! And I don't think "setup complexity" is an issue here, the whole point of toolchains is that a user doesn't have to muck around with JDKs when toolchains are in use. It's a one-off download if JDK 8 isn't already on the machine. After that things "just work".

Toolchains are prominently featured in the Kotlin docs for configuring a Gradle project so I don't think this config will surprise anyone.

It's not hard to remove use of the toolchain here though if it's not wanted.

Comment (szpak) [reply build.gradle.kts:103]: @3flex What is a "problem" with using the current approach with building the project (by default) with the JDK that is used to start Gradle?

Support for building with JDK 8 is [deprecated](https://github.com/gradle-nexus/publish-plugin/pull/303), so I don't know, if forcing users to have (or automatically download JDK 8) matches the aforementioned.

Comment (3flex) [reply build.gradle.kts:103]: It's not a problem. It just makes the config simpler and in line with best practice.

I'm going to remove the use of toolchains then to get this over the line.

File: build.gradle.kts (hunk 1 above)
Comment (3flex) [reply build.gradle.kts]: This was the last blocking bit of feedback which I believe I've answered above. Are you happy to approve?

**Review by TWiStErRob — APPROVED** [review]
